Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,381 @@
+        // FIXME: find only using the base characters i.e. also fold accents and others.
+        auto potentialIndex = searchBuffer.findIgnoringASCIICase(query, searchStart);
 
-    auto searchRange = makeRangeSelectingNodeContents(document);
+        if (potentialIndex == notFound)
+            return std::nullopt;
+        matchIndex = potentialIndex;
+        
+        start = boundaryPointAtIndexInNodes(*matchIndex, nodes, BoundaryPointIsAtEnd::No);
+        end = boundaryPointAtIndexInNodes(*matchIndex + query.length(), nodes, BoundaryPointIsAtEnd::Yes);
+        
+        if (((wordStartBounded == WordBounded::Yes) && !indexIsWordBoundary(searchBuffer, *matchIndex)) || ((wordEndBounded == WordBounded::Yes) && !indexIsWordBoundary(searchBuffer, *matchIndex + query.length()))) {
+            searchStart = *matchIndex + 1;
+            matchIndex = std::nullopt;
+        }
+    }
+        
+    unsigned endInset = 0;
+    if (nodes.last() == &searchRange.endContainer())
+        endInset = searchRange.endContainer().length() - searchRange.endOffset();
+    
+    if (*matchIndex + query.length() > searchBuffer.length() - endInset)
+        return std::nullopt;
+    
+    ASSERT_WITH_MESSAGE(start, "Scroll To Text Fragment: Start cannot be null");
+    ASSERT_WITH_MESSAGE(end, "Scroll To Text Fragment: End cannot be null");
+    
+    ASSERT_WITH_MESSAGE(contains(searchRange, start.value()), "Scroll To Text Fragment: Start is within SearchRange.");
+    ASSERT_WITH_MESSAGE(contains(searchRange, end.value()), "Scroll To Text Fragment: End is within SearchRange.");
+    
+    if (!start || !end || !contains(searchRange, start.value()) || !contains(searchRange, end.value()))
+        return std::nullopt;
     
-    FindOptions findOptions { DoNotRevealSelection };
+    return SimpleRange(*start, *end);
     
-    auto resultRange = collapseIfRootsDiffer(findPlainText(searchRange, target, findOptions));
+}
+// https://wicg.github.io/scroll-to-text-fragment/#find-a-string-in-range
+static std::optional<SimpleRange> rangeOfStringInRange(const String& query, SimpleRange& searchRange, WordBounded wordStartBounded, WordBounded wordEndBounded)
+{
+    while (!searchRange.collapsed()) {
+        Node* currentNode = &searchRange.startContainer();
+        
+        if (isNonSearchableSubtree(*currentNode)) {
+            if (auto newStart = nextSkippingChildrenInComposedTreeIgnoringUserAgentShadow(*currentNode))
+                searchRange.start = { *newStart, 0 };
+            else
+                return std::nullopt;
+            continue;
+        }
+        
+        if (!isVisibleTextNode(*currentNode)) {
+            do {
+                if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(*currentNode)) {
+                    searchRange.start = { *newStart, 0 };
+                    currentNode = newStart;
+                } else
+                    return std::nullopt;
+            } while (is<DocumentType>(currentNode));
```

Comment:
I think this will read better if we looped over DocType nodes and set range once instead.

Reply:
I struggled with this sections a lot, I do believe it needs to be done as a do-while to match the spec.

Reply:
ok